src/server/pfs/server/driver_v2.go

```diff
@@ -168,39 +165,87 @@ func (d *driverV2) withFileSet(ctx context.Context, repo, commit string, f func(
 	})
 }
 
-// TODO Need to figure out path cleaning.
-func (d *driverV2) getTar(ctx context.Context, repo, commit, glob string, w io.Writer) error {
-	if err := d.getTarConditional(ctx, repo, commit, glob, func(fr *FileReader) error {
-		return fr.Get(w, true)
-	}); err != nil {
+func (d *driverV2) getTar(ctx context.Context, commit *pfs.Commit, glob string, w io.Writer) error {
+	compactedPaths := []string{compactedCommitPath(commit)}
+	prefix := globLiteralPrefix(glob)
+	mf, err := matchFunc(glob)
+	if err != nil {
 		return err
 	}
-	// Close a tar writer to create tar EOF padding.
-	return tar.NewWriter(w).Close()
+	mr, err := d.storage.NewMergeReader(ctx, compactedPaths, index.WithPrefix(prefix))
+	if err != nil {
+		return err
+	}
+	filter := &fileset.IndexFilter{
+		R: mr,
+		F: func(idx *index.Index) bool {
+			return mf(idx.Path)
+		},
+	}
+	// TODO: remove absolute paths on the way out?
+	// nonAbsolute := &fileset.HeaderMapper{
+	// 	R: filter,
+	// 	F: func(th *tar.Header) *tar.Header {
+	// 		th.Name = "." + th.Name
+	// 		return th
+	// 	},
+	// }
+	return fileset.WriteTarStream(w, filter)
 }
 
-func (d *driverV2) listFileV2(pachClient *client.APIClient, file *pfs.File, full bool, history int64, f func(*pfs.FileInfoV2) error) error {
+func (d *driverV2) listFileV2(pachClient *client.APIClient, file *pfs.File, full bool, history int64, cb func(*pfs.FileInfoV2) error) error {
 	ctx := pachClient.Ctx()
 	if err := validateFile(file); err != nil {
 		return err
 	}
 	if err := d.checkIsAuthorized(pachClient, file.Commit.Repo, auth.Scope_READER); err != nil {
 		return err
 	}
-	// TODO: handle children of a directory
-	return d.getTarConditional(ctx, file.Commit.Repo.Name, file.Commit.ID, file.Path, func(fr *FileReader) error {
-		return f(fr.Info())
-	})
-}
-
-var globRegex = regexp.MustCompile(`[*?[\]{}!()@+^]`)
-
-func globLiteralPrefix(glob string) string {
-	idx := globRegex.FindStringIndex(glob)
-	if idx == nil {
-		return glob
+	name := strings.TrimRight(file.Path, "/")
+	// exact match
+	getExact := func() (fileset.FileSource, error) {
```

Comment:
I still think the two passes approach is unnecessary and it does not seem cleaner. I think you could get the same behavior with an index prefix of `name` and the following filter:
```
if idx.Path == name {
  return true
}
if strings.HasPrefix(idx.Path, name+"/") && idx.Path != name+"/" {
  return true
}
return false
```

Also, it seems like it could make sense to just have the index filter be a functional option in the index package that would get passed in the same way we pass in index options to the merge reader.

Reply:
This doesn't really do 2 passes. The exact match is hardly a pass. It's either 1 or none.  Using a prefix of `name` means we will also have to iterate and filter over things with prefix `name + x`.  Using a prefix of `name + "/"` means we only iterate over things in a hypothetical directory `name`, instead of all directories starting with `name`.

Reply:
I can add a Concat reader in the fileset package if you would rather only create one `server.Reader`

Reply:
Yeah, I don't think the performance gain in that very specific situation is worth the extra code. Also, I tweaked the filter a bit to account for the children of children stuff. My issue with the two passes approach is not so much about performance, since they will probably result with cache hits, I am more concerned with all of the extra code that has to be understood and maintained.